tests/utils_for_testbook.py

```diff
@@ -62,6 +63,49 @@ def inner_decorator(func: Callable) -> Any:
     return inner_decorator
 
 
+def _read_qmod_files(file_path: str):
+    return {
+        str(qmod_path): qmod_path.read_text()
+        for qmod_path in Path(file_path).parent.glob("*.qmod")
+    }
+
+
+def _normalize_qmod_code(code: str) -> str:
+    code = code.strip()
+    float_pattern = r"[-+]?(\d+(\.\d*)?|\.\d+)([eE][-+]?\d+)?"
+    code = re.sub(rf"\({float_pattern}\)", "<NUMBER>", code)
+    code = re.sub(rf"(?<=\W){float_pattern}", "<NUMBER>", code)
+    code = re.sub(r"\s+", "", code)
+    code = re.sub(r"Pauli::[IXYZ]", "<PAULI>", code)
+    return code
+
+
+def _compare_qmods(qmods: dict[str, str], new_qmods: dict[str, str]) -> None:
+    if len(new_qmods) > len(qmods):
+        uncommitted_files = [file for file in new_qmods if file not in qmods]
+        raise AssertionError(
+            f"Found uncommitted Qmod files: {', '.join(uncommitted_files)}"
+        )
+    for file in qmods:
+        assert _normalize_qmod_code(qmods[file]) == _normalize_qmod_code(
+            new_qmods[file]
+        ), f"Qmod file {file.split('/')[-1]} is not up-to-date"
+
+
+def _build_testbook_decorator(notebook_path: str, timeout_seconds: float) -> Callable:
+    def inner(func: Callable) -> Any:
+        def inner(*args: Any, **kwargs: Any) -> Any:
+            qmods = _read_qmod_files(notebook_path)
+            with testbook(notebook_path, execute=True, timeout=timeout_seconds) as tb:
+                new_qmods = _read_qmod_files(notebook_path)
+                _compare_qmods(qmods, new_qmods)
+                return func(tb, *args, *kwargs)
```

Comment:
The ordering here means that the Qmod comparison runs before the test's logic? Can this be inverted with the change below?

I'm not sure what order is better. For PRs I'd say it's better to first actually test the notebook and only then fail on the Qmod, but for the weekly runs I want to know exactly what Qmods are not up-to-date, even for notebooks that fail its test

```suggestion
                ret_value = func(tb, *args, *kwargs)
                new_qmods = _read_qmod_files(notebook_path)
                _compare_qmods(qmods, new_qmods)
                return ret_value
```

Reply:
Hmm... Yeah, I had the latter scenario in mind. In the weekly run, the Qmod comparison must run first so it would be the first failure.

I don't think that running it first on PRs is too bad... I think that the most damage is in local runs though, where the programmer would have to run the test twice each time. The development workflow would become change notebook -> run test -> fail on Qmod (but the test updates the Qmiod) -> fail on bug -> repeat.

Maybe we should enable the Qmod comparison only on the CI?... Or is it not too bad in your opinion?...